src/blockdata/script.rs

```diff
@@ -643,107 +678,339 @@ impl Script {
+            .push_opcode(opcodes::all::OP_HASH160)
+            .push_slice(&self.as_bytes()[2..]) // The `self` script is 0x00, 0x14, <pubkey_hash>
+            .push_opcode(opcodes::all::OP_EQUALVERIFY)
+            .push_opcode(opcodes::all::OP_CHECKSIG)
+            .into_script();
+
+        Some(script)
     }
 
     /// Creates an assembly decoding of the script in the given byte slice.
     pub fn bytes_to_asm(script: &[u8]) -> String {
         let mut buf = String::new();
-        Script::bytes_to_asm_fmt(script, &mut buf).unwrap();
+        bytes_to_asm_fmt(script, &mut buf).unwrap();
         buf
     }
 
-    /// Returns the assembly decoding of the script.
-    pub fn asm(&self) -> String {
-        Script::bytes_to_asm(self.as_ref())
+    /// Adds a single opcode to the script.
+    pub fn push_opcode(&mut self, data: opcodes::All) {
+        self.0.push(data.to_u8());
+    }
+
+    /// Adds instructions to push some arbitrary data onto the stack.
+    pub fn push_slice(&mut self, data: &[u8]) {
+        // Start with a PUSH opcode
+        match data.len() as u64 {
+            n if n < opcodes::Ordinary::OP_PUSHDATA1 as u64 => { self.0.push(n as u8); },
+            n if n < 0x100 => {
+                self.0.push(opcodes::Ordinary::OP_PUSHDATA1.to_u8());
+                self.0.push(n as u8);
+            },
+            n if n < 0x10000 => {
+                self.0.push(opcodes::Ordinary::OP_PUSHDATA2.to_u8());
+                self.0.push((n % 0x100) as u8);
+                self.0.push((n / 0x100) as u8);
+            },
+            n if n < 0x100000000 => {
+                self.0.push(opcodes::Ordinary::OP_PUSHDATA4.to_u8());
+                self.0.push((n % 0x100) as u8);
+                self.0.push(((n / 0x100) % 0x100) as u8);
+                self.0.push(((n / 0x10000) % 0x100) as u8);
+                self.0.push((n / 0x1000000) as u8);
+            }
+            _ => panic!("tried to put a 4bn+ sized object into a script!")
+        }
+        // Then push the raw bytes
+        self.0.extend_from_slice(data);
+    }
+
+    /// Adds an `OP_VERIFY` to the script, unless the most-recently-added
+    /// opcode has an alternate `VERIFY` form, in which case that opcode
+    /// is replaced e.g., `OP_CHECKSIG` will become `OP_CHECKSIGVERIFY`.
+    ///
+    /// This function needs to iterate over the script to find the last instruction. Prefer
+    /// `Builder` if you're creating the script from scratch or if you want to push `OP_VERIFY`
+    /// multiple times.
+    pub fn scan_and_push_verify(&mut self) {
+        self.push_verify(self.last_opcode());
+    }
+
+    fn push_verify(&mut self, last_opcode: Option<opcodes::All>) {
+        match opcode_to_verify(last_opcode) {
+            Some(opcode) => {
+                self.0.pop();
+                self.push_opcode(opcode);
+            },
+            None => self.push_opcode(opcodes::all::OP_VERIFY),
+        }
+    }
+}
+
+impl Deref for ScriptBuf {
```

Comment:
My mind was just blown, this is acting like compile time polymorphism, right? This means that any method on `Script` can be called on a `ScriptBuf` because the compiler automatically dereferences the `ScriptBuf`, that is hot AF.

Reply:
Not really polymorphism, just `Deref` coercions. They work very well in case of `String` or `Vec` in my experience, so I went with it.

Reply:
I wouldn't use the word "just" to refer to `Deref` coercions :). I agree that this is one of the weirdest most magical parts of the language.